### PR TITLE
DBZ-1385 eventType can be configured as add field

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-1385

Current implementation expects that evetn type is mapped to field named `eventType`. If it is the case the configuration is completely taken from additional fields. If such add field does not exist then `eventType` field is added to payload.

Complete backward compatibility is ensured. `eventType` field is always first in the message struct.

### CODE REVIEW

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java [hunk 1]
```diff
@@ -255,18 +251,29 @@ private Schema getValueSchema(Schema debeziumEventSchema, Integer version) {
     private SchemaBuilder getSchemaBuilder(Schema debeziumEventSchema) {
         SchemaBuilder schemaBuilder = SchemaBuilder.struct();
 
+        // eventType is first from due to backward compatibility reasons
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: Didn't we conclude that sole re-ordering isn't a problem in terms of backwards-compatibility? That said, eventType used to come before payload, whereas now it doesn't any longer, so isn't there an ordering change in any case?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: @gunnarmorling Yes we did but it felt to me this is cheap price to keep it same.

**Comment by jpechane** [comment]
@gunnarmorling Regarding the promotion - I kept this out of this change. We should decide if the promotion should be automatic or not.

File: debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: @jpechane, but the order still *is* changed, no? So it seems the current approach adds some complexity for not much gain.

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: @gunnarmorling No with current approach the order is kept

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: How so? It used to come *before* the payload, and now it's *after* that, no?

Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: @gunnarmorling No, it first finds eventType, then adds payload and then it add the rest of additional fields

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/transforms/outbox/EventRouter.java]: Ah yes, I stand corrected. But taking a step back -- the idea was to *simplify* things and make them more regular, I don't feel this is what's happening here ;) Also keep in mind the feature is marked as "incubating" in the docs, so I don't think we should bother too much for keeping that ordering here, esp. as it doesn't have any implications on compatibility anyways.

Another question: it seems there's no way to *not* have the event type field in the messages, or is there? Tbh. I think I'd just remove any notion of "event type" from the implementation altogether. After all, it's not an inherent detail of the pattern (unlike the aggregate type which definitely is needed for topic routing). So if the user wants to have it, they can just add it themselves as an additional field.

**Comment by gunnarmorling** [comment]
>  We should decide if the promotion should be automatic or not.

Yes, I think it should be done automatically. No need for nesting if there's a single element. Also will be more efficient, as the payload contents can be passed through as-is in that (common) case.

**Comment by jpechane** [comment]
@gunnarmorling Added payload promotion

**Comment by gunnarmorling** [comment]
@hpgrahsl, thought this one might be interesting for you, too. The outbox format will slightly change: payload promoted to be the sole message value if there's no other "in-message" attribute, and "eventType" is no special attribute any longer, but just can be propagated by the user as an "additional attribute" in the header or message value.

**Comment by gunnarmorling** [comment]
@jpechane, rebased and applied. Thanks!

**Comment by renatomefi** [comment]
Hey, sorry for the delay into checking this one!

I personally prefer explicit vs implicit behavior, if a user adds a field to the payload they'll find a completely different format, I think a new configuration could ensure users won't have this sort of implicit changes.

Also the thing about expanding the payload instead of serializing it has been discussed before right? Would be nice to make it possible independently of the additional fields.

What do you think about it?

**Comment by gunnarmorling** [comment]
> I personally prefer explicit vs implicit behavior, if a user adds a field to the payload they'll find a completely different format, I think a new configuration could ensure users won't have this sort of implicit changes.

The thing is, I'm not a fan of adding more and more options, the outbox SMT is already very complex to use. What would be the advantage of nesting the message value, if there's just a single field which goes into it? Note I'd assume this to be the default case, and other fields going to the header most of the times. In that case, the nesting also poses a performance penalty, as otherwise the value just can taken from the payload column as is, without the need for further modifications to the `struct`.

> Also the thing about expanding the payload instead of serializing it has been discussed before right? Would be nice to make it possible independently of the additional fields.

Yes, I'd love to see that one, too. Someone from the community has began to work on it, but it seems to have stalled a bit. Would be very nice to have this for sure :)

**Pull request merged**